posixlib/src/main/scala/scala/scalanative/runtime/SocketHelpers.scala

```diff
@@ -156,67 +152,63 @@ object SocketHelpers {
       if (status != 0)
         return scala.Array.empty[String]
 
-      var p = !ret
-      while (p != null) {
-        var ipstr = alloc[CChar]((INET6_ADDRSTRLEN + 1).toUInt)
-        var addr  = alloc[Byte]
-        if (p.ai_family == AF_INET) {
-          addr = p.ai_addr
-            .asInstanceOf[Ptr[sockaddr_in]]
-            .sin_addr
-            .toPtr
-            .asInstanceOf[Ptr[Byte]]
-        } else {
-          addr = p.ai_addr
-            .asInstanceOf[Ptr[sockaddr_in6]]
-            .sin6_addr
-            .toPtr
-            .asInstanceOf[Ptr[Byte]]
-        }
-        inet_ntop(p.ai_family, addr, ipstr, INET6_ADDRSTRLEN.toUInt)
+      var ai = !ret
+      while (ai != null) {
+        val ipstr = stackalloc[CChar]((INET6_ADDRSTRLEN + 1).toUInt)
+        val addr =
+          if (ai.ai_family == AF_INET) {
+            ai.ai_addr
+              .asInstanceOf[Ptr[sockaddr_in]]
+              .sin_addr
+              .toPtr
+              .asInstanceOf[Ptr[Byte]]
+          } else {
+            ai.ai_addr
+              .asInstanceOf[Ptr[sockaddr_in6]]
+              .sin6_addr
+              .toPtr
+              .asInstanceOf[Ptr[Byte]]
+          }
+        inet_ntop(ai.ai_family, addr, ipstr, INET6_ADDRSTRLEN.toUInt)
         retArray += fromCString(ipstr)
-        p = p.ai_next.asInstanceOf[Ptr[addrinfo]]
+        ai = ai.ai_next.asInstanceOf[Ptr[addrinfo]]
       }
-      freeaddrinfo(!ret)
+      freeaddrinfo(!ret) // start from first addrinfo
       retArray.toArray
     }
-  }
 
-  def ipToHost(ip: String, isV6: Boolean): Option[String] = {
+  def ipToHost(ip: String, isV6: Boolean): Option[String] =
     Zone { implicit z =>
-      var status  = 0
-      val host    = alloc[CChar](1024.toUInt)
-      val service = alloc[CChar](20.toUInt)
-      if (isV6) {
-        val addr6 = alloc[sockaddr_in6]
-        addr6.sin6_family = AF_INET6.toUShort
-        inet_pton(AF_INET6,
-                  toCString(ip),
-                  addr6.sin6_addr.toPtr.asInstanceOf[Ptr[Byte]])
-        status = getnameinfo(addr6.asInstanceOf[Ptr[sockaddr]],
-                             sizeof[sockaddr_in6].toUInt,
-                             host,
-                             1024.toUInt,
-                             service,
-                             20.toUInt,
-                             0)
-      } else {
-        val addr4 = alloc[sockaddr_in]
-        addr4.sin_family = AF_INET.toUShort
-        inet_pton(AF_INET,
-                  toCString(ip),
-                  addr4.sin_addr.toPtr.asInstanceOf[Ptr[Byte]])
-        status = getnameinfo(addr4.asInstanceOf[Ptr[sockaddr]],
-                             sizeof[sockaddr_in].toUInt,
-                             host,
-                             1024.toUInt,
-                             service,
-                             20.toUInt,
-                             0)
-      }
-
+      val host    = stackalloc[CChar](1024.toUInt)
```

Comment:
I advise using `alloc` here. `stackalloc` works, but carries a potential
downside.

Allocating 1KiB on the stack, especially in the context of other stack allocations
is likely to lead to "out of stack" issues.

In single thread programs, such as SN now, this would manifest by failure far down a long call chain. By nature, those are likely to be intermittent and hard to find.  Once discovered, the fix is trivial: use alloc.

Al there are currently no, or very few, multi-thread SN programs, one must 
think and design ahead.  Threaded programs tend to have smaller, and often quite limited, stacks.

Reply:
Why was 1024 chosen?

Anyway, in the docs that include 500 bytes, everything is done on the stack - that was my model - plus aren't the lifetimes on the stack very short lived?

https://man7.org/linux/man-pages/man3/getaddrinfo.3.html

Looking at this - if it is the kernel then it shouldn't be too big but maybe 1kb on the stack seems like it should be fine. Maybe we should consider greater that 1k iffy.
https://stackoverflow.com/questions/22882688/suggested-max-size-for-stack-allocations

Reply:
I found this too - https://stackoverflow.com/questions/40218878/should-the-host-name-array-be-declared-maxhostnamelen1-for-gethostname

Reply:
@ekrich in good will.,, Your PR, your judgement call.  

I offer the below as discussion and _do_not_ want to introduce
contention or, especially, to block or delay the merge of the PR.
Freedom in discussion/counsel, unity in decision/action. 

I can give this a LGTM with a note that says the 1KiB stackalloc
is not the way I would do it, but not dead wrong.

All I can do is say "Here be lions!".   I am speaking
from extensive experience implementing and supervising
the implementation of TCP/IP at both user & kernel levels.

1KiB may or may not be a lion, but it has four legs, is tawny,
has a mane, and roars like the lion at the beginning of many
movies. 

Again, I point out that any execution time savings is 10 or 20 orders
of magnitude less than the cost of the trip up and down the network
stack.

I think the PR has benefit and I do not want to hold it up when there are
far more important matters afoot.   Like fixing blatant
errors in stackalloc/alloc and getting the latter to 
use a tiered bump allocator so that we all escape
the discussion of "but alloc uses malloc on each and
every call and is so slow".   


re: `Why was 1024 chosen` First and best answer is: I do not know.  
second answer is "Looks like room enough for several addrinfos'.
Now, that rapidly takes me down the rathole that the code in question
is not robust.  It may work on small systems, but should probably
be doing a retry if there are more addresses than the buffer will 
allow.  Of course, one would need to trace the entire method
to see if only one or a few addr info were required.  
I have no need to venture into that Swamp and will whistle past it:.

re: `everything is done on the stack`  I suggest a decision rule of `everything not-largish is done on the stack`
where the value of not-largish is the point of the discussion.  Your PR, your call, said in good will.

re: `plus aren't the lifetimes on the stack very short lived?`  The lifefime of stack variables is the scope of the enclosing block.  That means the cumulative stack usage depends upon the depth and allocations of the entire call chain below that scope.  Do you know the stack usage of all the getaddrinfo and everything
it calls?  

Agreed that this is not kernel code, but it is library code.  Library code needs to be
non-profligate in stack usage also.

Reply:
I am ok with it now - it is just the host name size and now it is 256. I agree that we should be checking return values etc. to make this more robust. I think it is code going in the right direction. Certainly, as I read online even 256 might be to large for a kernel or driver but I think here in user code it should be fine. This of course is a record so will see how this pans out.